### ISSUE TITLE
3x times slower IO when all CPU cores are available

### ISSUE THREAD

**Issue by petr-ujezdsky** [docker/for-mac#3771]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: `028F3CF0-E3EF-4D96-8D0A-5C8B98A48CE7/20190730143332`

### Expected behavior
IO performance should be independent of number of available CPU cores.

### Actual behavior
Running the `dd` command takes 3x more time when all (8) cores are available in Docker Settings.

### Information

  - macOS Version: 10.14.6
  - MacBook Pro 15", 2017
  - 2,9 GHz Intel Core i7

### Diagnostic logs

```
Docker for Mac: Version 2.0.0.3 (31259)
```

### Steps to reproduce the behavior

  1. Configure `Docker for Mac` to use all available CPU cores
   Preferences -> Advanced -> CPUs -> Maximum (8 for me)
  2. Run following command in some temporary folder. It will create 100MB `test.dat` file here that you can then delete.
  ```
  $ docker run --rm -it -v $(pwd)/:/var/www alpine time dd if=/dev/zero of=/var/www/test.dat bs=1024 count=100000
100000+0 records in
100000+0 records out
real	1m 4.23s
user	0m 0.70s
sys	0m 9.50s
  ```
  3. Observe the long duration (over 1 minute)
  4. Go to step #1 and allow only 7 or less CPUs:
  ```$ docker run --rm -it -v $(pwd)/:/var/www alpine time dd if=/dev/zero of=/var/www/test.dat bs=1024 count=100000
100000+0 records in
100000+0 records out
real	0m 20.53s
user	0m 0.05s
sys	0m 2.87s
  ```
  5. Now you can see the difference in performance (3 times faster)

I have also tried to restrict number of CPUs per container but this seems to have no effect (still 3x slower on all 8 cores available).
```
docker run --cpus=1 ...
```
I found out about this issue thanks to `docker-sync` performance site (https://docker-sync.readthedocs.io/en/latest/miscellaneous/performance.html) where this command is used as benchmark.

**Comment by petr-ujezdsky** [comment]
I have created more thorough IO performance test on my github - https://github.com/petr-ujezdsky/docker-io-test.
It contains data and graphs.

**Comment by petr-ujezdsky** [comment]
I have just updated to newer **version `2.1.0.0`** and the **problem still persists**.

Test times are +- identical to the previous 2.0.0.3 version.

**Comment by petr-ujezdsky** [comment]
Version 2.1.0.1, still the same problem
Diagnostics ID: 028F3CF0-E3EF-4D96-8D0A-5C8B98A48CE7/20190809165210

**Comment by petr-ujezdsky** [comment]
Version 2.1.0.3, still relevant. It is actually **worse by 30% compared to v 2.0.0.3**.
Diagnostics ID: 028F3CF0-E3EF-4D96-8D0A-5C8B98A48CE7/20191024200015

**Comment by djs55** [comment]
Thanks for the report. I'll reopen the issue.

/remove-lifecycle stale

**Comment by petr-ujezdsky** [comment]
Ok so finally I've had free time today and I have run tests for `v2.3.0.2`.
**And it actually seems to be fixed by #4423!**

![docker-io-performance-overview](https://user-images.githubusercontent.com/2472013/81800924-e4935700-9513-11ea-8fdf-c442b1a89625.png)

I have updated my [git repo](https://github.com/petr-ujezdsky/docker-io-test) with these better graphs.

The performance still seems to be a little bit dependent on the available core count. But I think that might be the CPU scheduler thing, since the IO is utilising CPU so much.

Now I only wish the IO performance would be closer to native or at least not 40x worse with very high CPU utilisation.. :)

Thank you very much for your work, I highly appreciate that!
